File: packaging/sdk/recipes/c05_autogen/check.py
"""Run the frozen Phase 3 reference workload through the AgentChat application.

The instrumentation the recipe adds: wrap each agent's model client in
``ZerothChatCompletionClient`` with the model and provider it bills, open a run
per request, charge tools you price yourself, record the outcome you own.
"""

from __future__ import annotations

import argparse
import asyncio
import importlib.util
import json
import sys
from datetime import timedelta
from decimal import Decimal
from pathlib import Path

from zeroth.instrumentation import Recorder
from zeroth.instrumentation.autogen import ZerothChatCompletionClient


def _load(name: str, path: Path):
    spec = importlib.util.spec_from_file_location(name, path)
    module = importlib.util.module_from_spec(spec)
    sys.modules[name] = module
    spec.loader.exec_module(module)
    return module


app = _load("c05_autogen_app", Path(__file__).with_name("app.py"))
fakes = _load("c05_autogen_fakes", Path(__file__).with_name("fakes.py"))


def wrap(clients: fakes.Clients, model: str, step: str) -> ZerothChatCompletionClient:
    provider = "anthropic" if model.startswith("claude") else "openai"
    return ZerothChatCompletionClient(
        clients.fake(model), model=model, provider=provider, step=step
    )


async def drive(clients: fakes.Clients, run, spec: dict, version: str) -> None:
    calls, question = spec["calls"], f"question for {spec['id']}"

    def queued(index: int, text: str = "done", **overrides):
        call = calls[index]
        clients.fake(call["model"]).expect(fakes.reply(text, call["usage"][version], **overrides))

    family = spec["id"]
    if family == "success":
        queued(0, "plan"), queued(1, "TERMINATE done")
        await app.plan_then_write(
            wrap(clients, app.PLANNER, "planner"), wrap(clients, app.WRITER, "writer"), question
        )
    elif family == "rejection":
        queued(0)
        await app.run_single("reviewer", wrap(clients, app.REVIEWER, "reviewer"), question)
    elif family in ("delayed_outcome", "duplicate_delivery"):
        queued(0)
        await app.run_single("quick", wrap(clients, app.QUICK, "quick"), question)
    elif family == "retry":
        queued(0, "not json"), queued(1, '{"ok": true}')
        await app.robust_extract(wrap(clients, app.PLANNER, "planner"), question)
    elif family == "parallel_calls":
        queued(0), queued(1), queued(2)
        planner = wrap(clients, app.PLANNER, "planner")
        await app.fanout(planner, [question + " a", question + " b"])
        await app.run_single("quick", wrap(clients, app.QUICK, "quick"), "merge")
    elif family == "tool_cost":
        tool = calls[1]
        run.tool_charge(
            tool["step"], tool=tool["tool"], cost_usd=Decimal(tool["cost_usd"][version])
        )
        queued(0)
        await app.run_single("reviewer", wrap(clients, app.REVIEWER, "reviewer"), question)
    elif family == "fallback":
        clients.fake(calls[0]["model"]).expect(RuntimeError("provider unavailable"))
        queued(1)
        await app.answer_with_fallback(
            wrap(clients, app.WRITER, "writer"), wrap(clients, app.REVIEWER, "reviewer"), question
        )
    elif family == "missing_usage":
        queued(0)
        await app.run_single("planner", wrap(clients, app.PLANNER, "planner"), question)
    else:
        raise ValueError(f"unmapped family {family}")


async def reference(recorder: Recorder, version: str, workload) -> fakes.Clients:
    clients = fakes.Clients()
    for spec in workload.RUNS:
        at = workload.run_time(spec["id"])
        with recorder.run(workload.WORKFLOW, version, spec["id"]).active() as run:
            await drive(clients, run, spec, version)
            run.summary(
                spec["terminal"], recorded_at=at, metadata={"framework": "autogen-agentchat"}
            )
            if spec["outcome"]["delayed"]:
                run.outcome(None, maturity="provisional", occurred_at=at)
                at = at + timedelta(hours=1)
            run.outcome(spec["outcome"]["accepted"], occurred_at=at)
    assert clients.pending == 0, "every queued reply was consumed"
    return clients


async def modes(recorder: Recorder, workflow: str) -> dict:
    """Mode scenarios; the harness asserts their rows from the database."""
    from autogen_core import CancellationToken

    clients = fakes.Clients()
    report: dict = {}
    with recorder.run(workflow, "v1", "modes-team").active() as run:
        clients.fake(app.PLANNER).expect(fakes.reply("plan", [100, 10]))
        clients.fake(app.WRITER).expect(fakes.reply("TERMINATE done", [200, 20]))
        report["team"] = await app.plan_then_write(
            wrap(clients, app.PLANNER, "planner"), wrap(clients, app.WRITER, "writer"), "q"
        )
        run.summary("completed")
        run.outcome(True)
    with recorder.run(workflow, "v1", "modes-stream").active():
        clients.fake(app.QUICK).expect(fakes.reply("streamed reply", [30, 3]))
        text, chunks = await app.stream_single("quick", wrap(clients, app.QUICK, "quick"), "q")
        report["stream"] = {"text": text, "chunks": chunks}
    with recorder.run(workflow, "v1", "modes-missing").active():
        clients.fake(app.PLANNER).expect(fakes.reply("no usage", None))
        await app.run_single("planner", wrap(clients, app.PLANNER, "planner"), "q")
    with recorder.run(workflow, "v1", "modes-cached").active():
        clients.fake(app.PLANNER).expect(fakes.reply("from cache", [40, 4], cached=True))
        await app.run_single("planner", wrap(clients, app.PLANNER, "planner"), "q")
    with recorder.run(workflow, "v1", "modes-cancelled").active():
        token = CancellationToken()
        token.cancel()
        client = wrap(clients, app.PLANNER, "planner")
        clients.fake(app.PLANNER).expect(fakes.reply("never", [1, 1]))
        try:
            await client.create([], cancellation_token=token)
        except RuntimeError as error:
            report["cancelled"] = str(error)
        clients.fake(app.PLANNER).queue.clear()
    assert clients.pending == 0
    return report


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--zeroth-url", required=True)
    parser.add_argument("--zeroth-key", required=True)
    parser.add_argument("--workload", required=True)
    parser.add_argument("--version", default="v1")
    parser.add_argument("--scenario", choices=["reference", "modes"], default="reference")
    args = parser.parse_args(argv)
    import autogen_agentchat
    import autogen_core
    from zeroth.sdk import ZerothClient

    workload = _load("reference_workload", Path(args.workload))
    recorder = Recorder(ZerothClient(api_key=args.zeroth_key, base_url=args.zeroth_url))
    report: dict = {}
    if args.scenario == "reference":
        asyncio.run(reference(recorder, args.version, workload))
    else:
        report = asyncio.run(modes(recorder, workload.WORKFLOW))
    print(
        json.dumps(
            {
                "python": sys.version.split()[0],
                "autogen_agentchat": autogen_agentchat.__version__,
                "autogen_core": autogen_core.__version__,
                "lost": len(recorder.lost),
                "version": args.version,
                "scenario": args.scenario,
                "report": report,
            }
        )
    )
    return 0 if not recorder.lost else 1


if __name__ == "__main__":
    raise SystemExit(main())

File: packaging/sdk/recipes/c06_vercel/check.ts
// Run the frozen Phase 3 reference workload and the C06 mode scenarios through the AI SDK app.
//
//   node check.ts --zeroth-url URL --zeroth-key KEY --workload workload.json [--version v1]
//                 [--scenario reference|modes] [--workflow NAME] [--retries N] [--timeout-ms MS]
import { parseArgs } from "node:util";
import { readFile } from "node:fs/promises";
import { Capture } from "./capture.ts";
import { Replies } from "./mocks.ts";
import * as app from "./app.ts";

type Call = { step: string; attempt: number; kind: string; model?: string; usage?: Record<string, [number, number, number] | null>; tool?: string; cost_usd?: Record<string, string> };
type RunSpec = { id: string; terminal: string; outcome: { accepted: boolean; delayed: boolean }; recorded_at: string; calls: Call[] };

const { values } = parseArgs({ options: {
  "zeroth-url": { type: "string" }, "zeroth-key": { type: "string" }, workload: { type: "string" },
  version: { type: "string", default: "v1" }, scenario: { type: "string", default: "reference" },
  workflow: { type: "string" }, retries: { type: "string", default: "3" }, "timeout-ms": { type: "string", default: "5000" } } });
const url = values["zeroth-url"]!.replace(/\/$/, "");
const key = values["zeroth-key"]!;
const workload = JSON.parse(await readFile(values.workload!, "utf8")) as { workflow: string; runs: RunSpec[] };
const workflow = values.workflow ?? workload.workflow;
const version = values.version!;
const delivery = { retries: Number(values.retries), timeoutMs: Number(values["timeout-ms"]) };

function capture(runId: string, recordedAt?: string): Capture {
  return new Capture({ url, key, workflow, version, runId, recordedAt, ...delivery });
}

function queued(replies: Replies, call: Call, text = "ok") {
  const tokens = call.usage![version];
  replies.expect(call.model!, { text, usage: tokens === null ? null : tokens });
}

/** One model call the way the recipe instruments it: callbacks charge steps, call() charges failures. */
function traced(cap: Capture, step: string, model: string, fn: (hooks: Record<string, unknown>) => Promise<unknown>) {
  return cap.call(step, model, () => fn(cap.callbacks(step, model)));
}

async function reference(): Promise<Capture[]> {
  const replies = new Replies();
  const captures: Capture[] = [];
  for (const run of workload.runs) {
    const cap = capture(run.id, run.recorded_at);
    captures.push(cap);
    const q = `question for ${run.id}`;
    const [c0, c1, c2] = run.calls;
    switch (run.id) {
      case "success":
        queued(replies, c0); queued(replies, c1);
        await traced(cap, "plan", c0.model!, (h) => app.plan(replies.model(c0.model!), q, h));
        await traced(cap, "answer", c1.model!, (h) => app.plan(replies.model(c1.model!), q, h));
        break;
      case "rejection": case "delayed_outcome": case "duplicate_delivery": case "missing_usage":
        queued(replies, c0);
        await traced(cap, "answer", c0.model!, (h) => app.plan(replies.model(c0.model!), q, h));
        break;
      case "retry":
        queued(replies, c0, "not json"); queued(replies, c1, '{"ok": true}');
        await app.robustExtract(replies.model(c0.model!), q, () => cap.callbacks("answer", c0.model!));
        break;
      case "parallel_calls":
        queued(replies, c0); queued(replies, c1); queued(replies, c2);
        await app.fanout(replies.model(c0.model!), [`${q} a`, `${q} b`], () => cap.callbacks("fanout", c0.model!));
        await traced(cap, "merge", c2.model!, (h) => app.plan(replies.model(c2.model!), "merge", h));
        break;
      case "tool_cost":
        cap.tool(c1.step, c1.tool!, c1.cost_usd![version]);
        queued(replies, c0);
        await traced(cap, "answer", c0.model!, (h) => app.plan(replies.model(c0.model!), q, h));
        break;
      case "fallback":
        replies.expect(c0.model!, { usage: null, error: "provider unavailable" });
        queued(replies, c1);
        try {
          await traced(cap, "answer", c0.model!, (h) => app.plan(replies.model(c0.model!), q, h));
        } catch {
          await traced(cap, "answer", c1.model!, (h) => app.plan(replies.model(c1.model!), q, h));
        }
        break;
      default:
        throw new Error(`unmapped family ${run.id}`);
    }
    cap.summary(run.terminal);
    let at = run.recorded_at;
    if (run.outcome.delayed) {
      cap.outcome(null, "provisional", at);
      at = new Date(Date.parse(at) + 3_600_000).toISOString();
    }
    cap.outcome(run.outcome.accepted, "final", at);
  }
  if (replies.pending) throw new Error(`${replies.pending} queued replies were not consumed`);
  return captures;
}

async function modes(): Promise<{ captures: Capture[]; report: Record<string, unknown> }> {
  const replies = new Replies();
  const captures: Capture[] = [];
  const report: Record<string, unknown> = {};
  const open = (id: string) => { const cap = capture(id); captures.push(cap); return cap; };

  let cap = open("modes-multistep");
  replies.expect(app.PLANNER, { toolCall: { name: "web_search", input: { query: "q" } }, usage: [20, 5] }, { text: "answer", usage: [60, 9, 10] });
  const multi = await cap.call("answer", app.PLANNER, () => app.answerWithTool(replies.model(app.PLANNER), "q",
    cap.callbacks("answer", app.PLANNER), () => cap.tool("web_search", "web_search", "0.00500000")));
  cap.summary("completed", { total_usage: multi.totalUsage });
  cap.outcome(true);
  report.multistep = { text: multi.text, steps: multi.steps };

  cap = open("modes-stream");
  replies.expect(app.QUICK, { text: "streamed", usage: [10, 2] });
  report.stream = await app.streamAnswer(replies.model(app.QUICK), "q", cap.callbacks("answer", app.QUICK));
  cap.summary("completed");

  cap = open("modes-stream-abort");
  replies.expect(app.QUICK, { text: "streamed", usage: [11, 2] });
  const controller = new AbortController();
  const hooks = cap.callbacks("answer", app.QUICK);
  try {
    await app.streamAnswer(replies.model(app.QUICK), "q", hooks, controller.signal, 1).then(() => controller.abort());
  } catch (error) { report.streamAbortError = (error as Error).name; }
  await new Promise((r) => setTimeout(r, 200));
  cap.summary("aborted");

  cap = open("modes-timeout");
  replies.expect(app.REVIEWER, { text: "late", usage: [1, 1], delayMs: 800 });
  try {
    await cap.call("answer", app.REVIEWER, () => app.plan(replies.model(app.REVIEWER), "q", cap.callbacks("answer", app.REVIEWER), AbortSignal.timeout(50)));
  } catch (error) { report.timeoutError = (error as Error).name; }
  cap.summary("failed");

  cap = open("modes-unknown");
  replies.expect("vendor/unlisted-model", { text: "?", usage: [10, 10] });
  await traced(cap, "answer", "vendor/unlisted-model", (h) => app.plan(replies.model("vendor/unlisted-model"), "q", h));
  cap.summary("completed");

  report.providerRequests = replies.requests.length;
  return { captures, report };
}

const started = performance.now();
const { captures, report } = values.scenario === "modes" ? await modes() : { captures: await reference(), report: {} };
const reports = await Promise.all(captures.map((c) => c.flush()));  // flush: every delivery acknowledged before exit
const lost = reports.flatMap((r) => r.lost);
console.log(JSON.stringify({ node: process.versions.node, ai: JSON.parse(await readFile(new URL("./node_modules/ai/package.json", import.meta.url), "utf8")).version,
  scenario: values.scenario, version, workflow, delivered: reports.reduce((n, r) => n + r.delivered, 0),
  duplicates: reports.reduce((n, r) => n + r.duplicates, 0), retried: reports.reduce((n, r) => n + Object.values(r.attempts).filter((a) => a > 1).length, 0),
  lost, elapsed_ms: Math.round(performance.now() - started), report }));
process.exitCode = lost.length ? 1 : 0;
